### PR TITLE
Add CPU Support

### PR DESCRIPTION
**Pull request by qfournier**
# Summary

This pull request add support for CPU-based training and inference.

### Verification

The following checks were performed:

1. **Unit Tests**:  ran `python3 -m pytest` with all tests passing.
2. **Perplexity Evaluation**: ran `pseudo_ppl.ipynb` on CPU and GPU. The observed differences in perplexity are likely due to Flash Attention's known stability issues ([Is Flash Attention Stable?](https://arxiv.org/html/2405.02803v1)). These differences are most pronounced for longer proteins, such as `B3H5H9_ARATH`.
	- **GPU Results**:
	```csv
	>A6I415|A6I415_RAT,2.322495519944971
	>Q9NVV2|CS073_HUMAN,17.405731615848254
	>O00628|PEX7_HUMAN,2.0509484296749756
	>A0A8M9P7W8|A0A8M9P7W8_DANRE,8.219952027270358
	>A0A5G2R2Q0|A0A5G2R2Q0_PIG,2.1361886698034187
	>Q5SIM6|Q5SIM6_THET8,3.6456189770278944
	>M0RAH9|M0RAH9_RAT,1.8872439391766391
	>B3H5H9|B3H5H9_ARATH,11.666534732588502
	>F5H0U3|F5H0U3_HUMAN,2.5862320380231707
	>Q8IY34|S15A3_HUMAN,1.9109164906029565
	```
	- **CPU Results**:
	```csv
	>A6I415|A6I415_RAT,2.3576657751933143
	>Q9NVV2|CS073_HUMAN,17.405742716106037
	>O00628|PEX7_HUMAN,2.0509489355290564
	>A0A8M9P7W8|A0A8M9P7W8_DANRE,8.219951795613955
	>A0A5G2R2Q0|A0A5G2R2Q0_PIG,2.1361883295615334
	>Q5SIM6|Q5SIM6_THET8,3.645618198189389
	>M0RAH9|M0RAH9_RAT,1.9961777408179249
	>B3H5H9|B3H5H9_ARATH,14.365362470492723
	>F5H0U3|F5H0U3_HUMAN,2.586231461482032
	>Q8IY34|S15A3_HUMAN,1.8295168329427218
	```
3. **Training Test**: ran `pretrain.py` for 10 steps on CPU using the following command. Due to the significant runtime of CPU training, the loss was not compared to the baseline.
	```bash
	accelerate launch \
	--config_file=conf/accelerate_cpu.yaml \
	--num_processes=1 \
	--num_cpu_threads_per_process=32 \
	--gradient_clipping=1.0 \
	scripts/pretrain.py \
	hydra.run.dir=logs/AMPLIFY_350M \
	wandb.dir=logs/AMPLIFY_350M \
	wandb.name=AMPLIFY_350M \
	dataset=uniref100 \
	dataset.train.paths.uniref100=$SCRATCH/uniref100_train.csv \
	dataset.validation.paths.uniprot=$SCRATCH/uniprot_dev.csv \
	dataset.validation.paths.pdb=$SCRATCH/scop_dev.csv \
	dataset.validation.paths.oas=$SCRATCH/oas_dev.csv \
	trainer.dir=logs/AMPLIFY_350M \
	trainer.train.per_device_batch_size=8 \
	trainer.validation.per_device_batch_size=8 \
	trainer.gradient_accumulation_steps=1
	```

### CODE REVIEW

**Comment by qfournier** [comment]
As pointed out by @lolalebreton, a check should be implemented to ensure that the `pad_mask` is additive, as `xformers` does not support a multiplicative mask. I will include the necessary `assert` statement.

**Comment by qfournier** [comment]
@philippschw, the `scaled_dot_product_attention` and `memory_efficient_attention` functions do not have parameters, they only compute `softmax(QK^T / d**0.5) V` using the provided query (`Q`), key (`K`), and value (`V`) matrices, which are computed externally (see [`amplify.py`](https://github.com/chandar-lab/AMPLIFY/blob/98b0d713fd9d7ab2cbe34464d67a3ceb0ccf8cd9/src/amplify/model/amplify.py#L145)). As a result, AMPLIFY can be pre-trained on a GPU and then used on a CPU. The opposite is possible too, but pre-training on CPU would be atrociously slow.

**Comment by bschul02-amgen** [comment]
@qfournier - there's a small change needed to keep one of the tests passing (namely: `test__compare_sequences_to_out_of_sample__executes_as_expected`) due to what appears to be a hard-coded device string in the test case itself.

I can push up a patch to this branch, if you like?

**Comment by qfournier** [comment]
@bschul02-amgen Good catch! Yes, that would be helpful, thank you.
